src/sasa.h

```diff
@@ -0,0 +1,227 @@
+     */
+    double exposedArcLength(std::vector<std::pair<double, double>>& arcs) const;
+
+  public:
+    /**
+     * @brief updates sasa of target particles
+     * @param neighbours_data array of NeighbourData objects
+     * @param radii of the particles in the ParticleVector
+     * @param target_indices absolute indicies of target particles in ParticleVector
+     */
+    void updateSASA(const std::vector<SASA::NeighboursData>& neighbours_data, const std::vector<double>& radii,
+                    const std::vector<int>& target_indices);
+
+    /**
+     * @brief resizes areas buffer to size of ParticleVector
+     * @param space
+     */
+    void init(Space& spc);
+
+    /**
+     * @brief calculates neighbourData objects of particles specified by target indices in ParticleVector
+     * @param space
+     * @param target_indices absolute indicies of target particles in ParticleVector
+     */
+    const std::vector<SASA::NeighboursData> calcNeighbourData(Space& spc, const std::vector<int>& target_indices);
+
+    /**
+     * @brief calculates neighbourData object of a target particle specified by target indiex in ParticleVector
+     * @param space
+     * @param target_index indicex of target particle in ParticleVector
+     */
+    SASA::NeighboursData calcNeighbourDataOfParticle(Space& spc, const int target_index);
+
+    void update(Space& spc, const Change& change) {}
+
+    void sync(Space& other_space, const Change& change) {}
+
+    const std::vector<double>& getAreas() const { return areas; }
+
+    SASA(const json& j, Space& spc) : SASA(spc, j.value("radius", 1.4) * 1.0_angstrom, j.value("slices", 20)) {}
+    SASA(Space& spc, double probe_radius, int n_slices_per_atom)
+        : probe_radius(probe_radius), n_slices_per_atom(n_slices_per_atom) {}
+};
+
+// not used yet
+class SASACellList : public SASA {
+
+  private:
+    using CellList_t = CellList::CellListSpatial<CellList::CellListType<int, CellList::Grid::Grid3DPeriodic>>;
+    std::unique_ptr<CellList_t> cell_list; // pointer to cell list
+    double cell_size;                      // dimension of a single cell
+    using CellCoord = CellList::GridOf<CellList_t>::CellCoord;
+    std::vector<CellCoord> cell_offsets; // holds offsets which define a 3x3x3 cube around central cell
+
+  public:
+    SASACellList(const json& j, Space& spc)
+        : SASACellList(spc, j.value("radius", 1.4) * 1.0_angstrom, j.value("slices", 20)) {}
+
+    SASACellList(Space& spc, double probe_radius, int n_slices_per_atom) : SASA(spc, probe_radius, n_slices_per_atom) {}
+
+    void init(Space& spc) {
+
+        cell_offsets.clear();
+        for (auto i = -1; i <= 1; ++i) {
+            for (auto j = -1; j <= 1; ++j) {
+                for (auto k = -1; k <= 1; ++k) {
+                    cell_offsets.push_back(CellCoord(i, j, k));
+                }
+            }
+        }
+
+        double max_radius = 0.0;
+        for (const auto& particle : spc.particles) {
```

Comment:
here I would use `ranges::cpp20` like this:
``` cpp
auto diameters = particles | transform([](auto &particle){return particle.traits().sigma;})
auto max_radius = 0.5 * max_element(diameters);
```

Reply:
I put the radii buffer vector into SASABase, because I think it fits there more, therefore the max_radius in init() is done by ranges::max(radii)
I am unsure how to initialize a std::vector<double> with ranges. So I did it like this:
        std::for_each(spc.particles.begin(), spc.particles.end(),
                      [&](const Particle& particle) { radii.push_back(particle.traits().sigma * 0.5); });
        auto max_radius = ranges::max(radii);